scripts/changelog.js

```diff
@@ -1,44 +1,180 @@
+  By default this script will print the release notes to stdout.
 
-Ordinarily this is run via the gen-changelog shell script, which appends
-the result to the changelog.
+  [tag]
+  A tag to generate release notes for. Helpful for testing this script against
+  old releases. Leave this empty to look for the most recent tag.
 
-*/
+  [--write|-w]
+  When set it will update the changelog with the new release.
+  If a release with the same version already exists it will replace it, otherwise
+  it will prepend it to the file directly after the top level changelog title.
+
+  [--read|-r]
+  When set it will read the release notes for the tag from the CHANGELOG.md,
+  instead of fetching it. This is useful after release notes have been manually
+  edited and need to be pasted somewhere else.
+`
+
+const exec = (...args) => execSync(...args).toString().trim()
+
+const changelogTypes = new Map(
+  config.changelogTypes.filter(c => !c.hidden).map((c) => [c.type, c.section]))
+
+const logger = (init) => {
+  const i = 2
+  let indent = 0
+  const res = [init]
+  return {
+    toString () {
+      return res.join('\n').trim()
+    },
+    group (v) {
+      this.log(v)
+      indent += i
+    },
+    groupEnd () {
+      indent -= i
+    },
+    log (v) {
+      if (!v) {
+        res.push('')
+      } else {
+        res.push(`${' '.repeat(indent)}${v}`)
+      }
+    },
+  }
+}
+
+const RELEASE = {
+  get h1 () {
+    return '# Changelog' + this.sep
+  },
+  sep: '\n\n',
+  heading: '## ',
+  // versions in titles must be prefixed with a v
+  versionRe: semver.src[11].replace('v?', 'v'),
+  version (v) {
+    return v.startsWith('v') ? v : `v${v}`
+  },
+  date (d) {
+    return `(${d || exec('date +%Y-%m-%d')})`
+  },
+  title (v, d) {
+    return `${this.heading}${this.version(v)} ${this.date(d)}`
+  },
+}
+
+const findRelease = (contents, version) => {
+  const escRegExp = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
+
+  const titleSrc = (v) => [
+    '^',
+    RELEASE.heading,
+    v ? escRegExp(v) : RELEASE.versionRe,
+    ' ',
+    escRegExp(RELEASE.date()).replace(/\d/g, '\\d'),
+    '$',
+  ].join('')
+
+  const releaseSrc = [
+    '(',
+    titleSrc(RELEASE.version(version)),
+    '[\\s\\S]*?',
+    RELEASE.sep,
+    ')',
+    titleSrc(),
+  ].join('')
+
+  const release = contents.match(new RegExp(releaseSrc, 'm')) || []
+  return (release[1] || '').trim()
+}
+
+const assertArgs = (args) => {
+  const current = exec(`git rev-parse --abbrev-ref HEAD`)
+  const unpushed = exec(`git log HEAD..origin/${args.branch}`).length > 0
+
+  if (!args.read && current !== args.branch || unpushed) {
+    throw new Error(`Unpushed commits or not on the right branch:\n${current}\n${unpushed}`)
+  }
+
+  if (args.help) {
+    console.log(usage())
+    process.exit(0)
+  }
+
+  return args
+}
 
 const parseArgs = (argv) => {
   const result = {
+    file: resolve(__dirname, '..', 'CHANGELOG.md'),
+    branch: 'latest',
```

Comment:
I think this was originally `origin/latest` in an attempt to solve the "my local git differs from github" problem.  It obviously doesn't, but perhaps there is a way for us to have this guard when building the changelog?

Reply:
This function calls `assertArgs` at the end which will throw an error if you're not on `latest` or you have unpushed commits. This made me realize we should also add a check for the opposite: remote has commits that you dont.